### ISSUE TITLE
RuntimeFixedProperties weirdness with SystemConfiguration.isPropertySet

### ISSUE THREAD

**Issue by ctubbsii** [apache/accumulo#3529]
I was thinking through some of the problems that could occur with a misuse of `AccumuloConfiguration.isPropertySet`, and it occurred to me that it might be possible for the `SystemConfiguration` class to report that some properties are set when they no longer are. This is due to the way we handle certain "fixed" properties that cannot be changed at runtime. We still allow these properties to be set, but they will only take effect on a restart.

For example, if `gc.port` is set in ZooKeeper at system startup, it will be added to the `RuntimeFixedProperties`. Calls to `SystemConfiguration.isPropertySet(Property.GC_PORT)` will return true, even if the property is deleted from ZooKeeper. But if it was not initially set (and used a default value), then was subsequently set, it will still return true. It will only return false if it was not set in ZK at startup, and remains so.

So, the affirmative result is not revealing the presence of the property in the current runtime, or the current value in ZK, separately, but the `OR` of those two.

This is not necessarily a bug, and so I will close this issue immediately after opening, because I don't know that there's anything to be done about it. After all, `isPropertySet` is an internal API only. But, I wanted to document this strangeness, for reference, if it becomes an issue in future.

**Comment by EdColeman** [comment]
It really seems that isPropertySet is checking if the key exists, not necessarily that the property was set.  So far, it looks like any property that we have defined will return true - the check will climb the hierarchy and if not found, it will check the default - which will exist and return true.  This is even if the value is null or an empty string as the default value.

The only time that isPropertySet would seem to work as excepted would be for custom properties - if a custom property is set, it will return true.  If not, there will not be a default, so it will be false.  That still leaves the determination if the set value is valid or not up to the caller.

**Comment by ctubbsii** [comment]
> It really seems that isPropertySet is checking if the key exists, not necessarily that the property was set. So far, it looks like any property that we have defined will return true - the check will climb the hierarchy and if not found, it will check the default - which will exist and return true. This is even if the value is null or an empty string as the default value.

No, the DefaultConfiguration returns false, not true. It only looks like it's checking of the key exists in the other implementations, because the presence of any key in those implementations means a user has configured it... the content of all those other implementations are empty by default, so any presence of a key there means the user was involved. But, the DefaultConfiguration always returns false at the bottom (or top? depending on your preferred orientation) of the hierarchy.

> 
> The only time that isPropertySet would seem to work as excepted would be for custom properties - if a custom property is set, it will return true. If not, there will not be a default, so it will be false. That still leaves the determination if the set value is valid or not up to the caller.

No, it works fine for default values, too, because DefaultConfiguration.isPropertySet returns false, not true.

**Comment by ctubbsii** [comment]
Oh wait, for any fixed properties, there's always a value stored in RuntimeFixedProperties, so SystemConfiguration will always return `true` for those, even if they are using defaults. So, the situation I described initially is even simpler.... RuntimeFixedProperties always appear to be set, regardless of what the user has or hasn't done. Since the situation is actually simpler, I think we could probably store some extra information in RuntimeFixedProperties to easily detect the situation, and report the correct result. I'm going to open this issue and assign it to myself, because I think I can fix this behavior.

**Comment by ctubbsii** [comment]
To properly fix #3873's handling of deprecated properties, and detection of whether the new property is actually set by the user or not, we need to fix this first.
